kubernetes-maven-plugin - Skipped images still are pushed, which fails

### Description

As a dev I want to be able to skip the build and push of certain images based on property in my POM.

I can set up the build to skip certain images but it fails when being pushed as the images that were skipped do not exist. It should be possible to skip the deployment of specific images, via the same or a different configuration property.

Error:

> [ERROR] k8s: Error while trying to push the image: Unable to add tag [yyz.ocir.io/frxxxxxxxb/my-image:tf-014] to image [frxxxxxxxb/my-image:tf-014] : {"message":"No such image: frxxxxxxxb/my-image:tf-014"} (Not Found: 404) [Error while trying to push the image: Unable to add tag [yyz.ocir.io/frxxxxxxxb/my-image:tf-014] to image [frxxxxxxxb/my-image:tf-014] : {"message":"No such image: frxxxxxxxb/my-image:tf-014"} (Not Found: 404)]

### Info

* Eclipse JKube version : 1.5.1
* Maven version (`mvn -v`) : Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)

Configuration is as so:

```xml
<plugin>
    <groupId>org.eclipse.jkube</groupId>
    <artifactId>kubernetes-maven-plugin</artifactId>
    <version>1.5.1</version>
    <executions>
        <execution>
            <id>docker-build</id>
            <phase>package</phase>
            <goals>
                <goal>build</goal>
            </goals>
        </execution>
        <execution>
            <id>docker-registry-push</id>
            <phase>deploy</phase>
            <goals>
                <goal>push</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <images>
            <image>
                <name>${dockerRepoPrefix}/%a:%v</name>
                <alias>aws-deployer</alias>
                <build>
                    <dockerFile>${project.basedir}/Dockerfile</dockerFile>
                    <tags>
                        <tag>latest</tag>
                    </tags>
                    <args>
                        <terraform_version>${dkr-terraform-version}</terraform_version>
                    </args>
                </build>
            </image>
            <image>
                <name>${dockerRepoPrefix}/%a:tf-014</name>
                <alias>aws-deployer</alias>
                <build>
                    <dockerFile>${project.basedir}/Dockerfile</dockerFile>
                    <skip>${skipTfVers}</skip>
                    <args>
                        <terraform_version>0.14.11</terraform_version>
                    </args>
                </build>
            </image>
            <image>
                <name>${dockerRepoPrefix}/%a:tf-013</name>
                <alias>aws-deployer</alias>
                <build>
                    <dockerFile>${project.basedir}/Dockerfile</dockerFile>
                    <skip>${skipTfVers}</skip>
                    <args>
                        <terraform_version>0.13.7</terraform_version>
                    </args>
                </build>
            </image>
            <image>
                <name>${dockerRepoPrefix}/%a:tf-012</name>
                <alias>aws-deployer</alias>
                <build>
                    <dockerFile>${project.basedir}/Dockerfile</dockerFile>
                    <skip>${skipTfVers}</skip>
                    <args>
                        <terraform_version>0.12.26</terraform_version>
                    </args>
                </build>
            </image>
        </images>
    </configuration>
</plugin>
```

#### Repro Steps

Run the above config with the following command:
`mvn deploy -D'skipTfVers=true'`

Note that the build phase correctly skips the indicated images with the particular TF versions, but it fails when the deploy is due to occur as (with a clear local reg) the images do not exist. I could find no way to specify `skipPush` at the image level (hopefully I have just missed that)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -43,6 +43,7 @@ Usage:
 * Fix #1167: Replace apiextensions.k8s.io/v1beta1 by apiextensions.k8s.io/v1
 * Fix #1180: `k8s:watch` uses default namespace even if other namespace is configured
 * Fix #1190: OpenShiftBuildService doesn't apply resources in configured namespace
+* Fix #1195: Push goal doesn't respect skip option in image build configuration
 * Fix #1209: Remove WildFly Swarm support
 * Fix #1219: Bump kubernetes-client to 5.11.2
 * Fix #1213: SnakeYaml dependency from Kubernetes Client + uses SafeConstructor

--- a/jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/RegistryService.java
+++ b/jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/RegistryService.java
@@ -14,7 +14,6 @@
 package org.eclipse.jkube.kit.build.service.docker;
 
 import java.io.IOException;
-import java.util.Collection;
 
 import org.eclipse.jkube.kit.build.api.auth.AuthConfig;
 import org.eclipse.jkube.kit.build.service.docker.access.CreateImageOptions;
@@ -46,35 +45,33 @@ public class RegistryService {
     /**
      * Push a set of images to a registry
      *
-     * @param imageConfigs images to push (but only if they have a build configuration)
+     * @param imageConfig image to push (but only if they have a build configuration)
      * @param retries how often to retry
      * @param registryConfig a global registry configuration
      * @param skipTag flag to skip pushing tagged images
      * @throws IOException exception
      */
-    public void pushImages(Collection<ImageConfiguration> imageConfigs,
-                           int retries, RegistryConfig registryConfig, boolean skipTag) throws IOException {
-        for (ImageConfiguration imageConfig : imageConfigs) {
-            BuildConfiguration buildConfig = imageConfig.getBuildConfiguration();
-            String name = imageConfig.getName();
-            if (buildConfig != null) {
-                String configuredRegistry = EnvUtil.firstRegistryOf(
-                    new ImageName(imageConfig.getName()).getRegistry(),
-                    imageConfig.getRegistry(),
-                    registryConfig.getRegistry());
+    public void pushImage(ImageConfiguration imageConfig,
+                          int retries, RegistryConfig registryConfig, boolean skipTag) throws IOException {
+        BuildConfiguration buildConfig = imageConfig.getBuildConfiguration();
+        String name = imageConfig.getName();
+        if (buildConfig != null) {
+            String configuredRegistry = EnvUtil.firstRegistryOf(
+                new ImageName(imageConfig.getName()).getRegistry(),
+                imageConfig.getRegistry(),
+                registryConfig.getRegistry());
 
 
-                AuthConfig authConfig = createAuthConfig(true, new ImageName(name).getUser(), configuredRegistry, registryConfig);
+            AuthConfig authConfig = createAuthConfig(true, new ImageName(name).getUser(), configuredRegistry, registryConfig);
 
-                long start = System.currentTimeMillis();
-                docker.pushImage(name, authConfig, configuredRegistry, retries);
-                log.info("Pushed %s in %s", name, EnvUtil.formatDurationTill(start));
+            long start = System.currentTimeMillis();
+            docker.pushImage(name, authConfig, configuredRegistry, retries);
+            log.info("Pushed %s in %s", name, EnvUtil.formatDurationTill(start));
 
-                if (!skipTag) {
-                    for (String tag : imageConfig.getBuildConfiguration().getTags()) {
-                        if (tag != null) {
-                            docker.pushImage(new ImageName(name, tag).getFullName(), authConfig, configuredRegistry, retries);
-                        }
+            if (!skipTag) {
+                for (String tag : imageConfig.getBuildConfiguration().getTags()) {
+                    if (tag != null) {
+                        docker.pushImage(new ImageName(name, tag).getFullName(), authConfig, configuredRegistry, retries);
                     }
                 }
             }

--- a/jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/RegistryServiceTest.java
+++ b/jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/RegistryServiceTest.java
@@ -13,23 +13,25 @@
  */
 package org.eclipse.jkube.kit.build.service.docker;
 
-import java.io.IOException;
-import java.util.Collections;
-import java.util.Properties;
-
 import org.eclipse.jkube.kit.build.service.docker.access.CreateImageOptions;
 import org.eclipse.jkube.kit.build.service.docker.access.DockerAccess;
 import org.eclipse.jkube.kit.common.KitLogger;
 import org.eclipse.jkube.kit.common.RegistryConfig;
-
+import org.eclipse.jkube.kit.config.image.ImageConfiguration;
 import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
 import org.junit.Before;
 import org.junit.Test;
 import org.mockito.ArgumentCaptor;
 
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Properties;
+
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.Assert.assertThrows;
 import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
@@ -41,12 +43,23 @@ public class RegistryServiceTest {
   private DockerAccess dockerAccess;
   private QueryService queryService;
   private RegistryService registryService;
+  private ImageConfiguration imageConfiguration;
+  private RegistryConfig mockedRegistryConfig;
 
   @Before
   public void setUp() throws Exception {
+    imageConfiguration = ImageConfiguration.builder()
+        .name("foo/bar:0.0.1")
+        .build(BuildConfiguration.builder()
+            .from("basefoo/basebar:0.1.0")
+            .tags(Arrays.asList("latest", "0.0.1-slim"))
+            .build())
+        .build();
     dockerAccess = mock(DockerAccess.class);
     queryService = mock(QueryService.class);
+    mockedRegistryConfig = mock(RegistryConfig.class);
     registryService = new RegistryService(dockerAccess, queryService, new KitLogger.SilentLogger());
+    when(mockedRegistryConfig.getRegistry()).thenReturn("example.com");
   }
 
   @Test
@@ -99,5 +112,31 @@ public class RegistryServiceTest {
         eq("quay.io"), createImageOptionsCaptor.capture());
     assertThat(createImageOptionsCaptor.getValue().getOptions()).containsExactlyInAnyOrderEntriesOf(
         new CreateImageOptions().fromImage("quay.io/organization/image").tag("version").platform("linux/amd64").getOptions());
+  }
+
+  @Test
+  public void pushImage_whenValidImageConfigurationProvidedWithSkipTag_shouldNotPushAdditionalTags() throws IOException {
+    // When
+    registryService.pushImage(imageConfiguration, 1, mockedRegistryConfig, true);
+
+    // Then
+    verify(dockerAccess, times(1))
+        .pushImage(eq("foo/bar:0.0.1"), any(), eq("example.com"), anyInt());
+    verify(dockerAccess, times(1))
+        .pushImage(any(), any(), any(), anyInt());
+  }
+
+  @Test
+  public void pushImage_whenValidImageConfigurationProvided_shouldPushApplicableTags() throws IOException {
+    // When
+    registryService.pushImage(imageConfiguration, 1, mockedRegistryConfig, false);
+
+    // Then
+    verify(dockerAccess, times(1))
+        .pushImage(eq("foo/bar:0.0.1"), any(), eq("example.com"), anyInt());
+    verify(dockerAccess, times(1))
+        .pushImage(eq("foo/bar:latest"), any(), eq("example.com"), anyInt());
+    verify(dockerAccess, times(1))
+        .pushImage(eq("foo/bar:0.0.1-slim"), any(), eq("example.com"), anyInt());
   }
 }

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/AbstractImageBuildService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/AbstractImageBuildService.java
@@ -13,7 +13,10 @@
  */
 package org.eclipse.jkube.kit.config.service;
 
+import org.eclipse.jkube.kit.common.RegistryConfig;
 import org.eclipse.jkube.kit.config.image.ImageConfiguration;
+
+import java.util.Collection;
 
 public abstract class AbstractImageBuildService implements BuildService {
   private final JKubeServiceHub jKubeServiceHub;
@@ -24,14 +27,30 @@ public abstract class AbstractImageBuildService implements BuildService {
 
   protected abstract void buildSingleImage(ImageConfiguration imageConfiguration) throws JKubeServiceException;
 
+  protected abstract void pushSingleImage(ImageConfiguration imageConfiguration, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException;
+
   @Override
   public final void build(ImageConfiguration... imageConfigurations) throws JKubeServiceException {
+    processImage(this::buildSingleImage, "Skipped building", imageConfigurations);
+  }
+
+  @Override
+  public final void push(Collection<ImageConfiguration> imageConfigs, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException {
+    processImage(imageConfiguration -> pushSingleImage(imageConfiguration, retries, registryConfig, skipTag), "Skipped push", imageConfigs.toArray(new ImageConfiguration[0]));
+  }
+
+  @FunctionalInterface
+  private interface ImageConfigurationProcessor {
+    void process(ImageConfiguration imageConfiguration) throws JKubeServiceException;
+  }
+
+  private void processImage(ImageConfigurationProcessor imageConfigurationConsumer, String skipMessage, ImageConfiguration... imageConfigurations) throws JKubeServiceException {
     if (imageConfigurations != null) {
       for (ImageConfiguration imageConfiguration : imageConfigurations) {
         if (imageConfiguration.getBuildConfiguration() != null && imageConfiguration.getBuildConfiguration().getSkip()) {
-          jKubeServiceHub.getLog().info("%s : Skipped building", imageConfiguration.getDescription());
+          jKubeServiceHub.getLog().info("%s : %s", imageConfiguration.getDescription(), skipMessage);
         } else {
-          buildSingleImage(imageConfiguration);
+          imageConfigurationConsumer.process(imageConfiguration);
         }
       }
     }

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/DockerBuildService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/DockerBuildService.java
@@ -14,7 +14,6 @@
 package org.eclipse.jkube.kit.config.service.kubernetes;
 
 import java.io.IOException;
-import java.util.Collection;
 import java.util.Objects;
 
 import org.eclipse.jkube.kit.build.service.docker.DockerServiceHub;
@@ -67,9 +66,9 @@ public class DockerBuildService extends AbstractImageBuildService {
     }
 
     @Override
-    public void push(Collection<ImageConfiguration> imageConfigs, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException {
+    protected void pushSingleImage(ImageConfiguration imageConfiguration, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException {
         try {
-            dockerServices.getRegistryService().pushImages(imageConfigs, retries, registryConfig, skipTag);
+            dockerServices.getRegistryService().pushImage(imageConfiguration, retries, registryConfig, skipTag);
         } catch (IOException ex) {
             throw new JKubeServiceException("Error while trying to push the image: " + ex.getMessage(), ex);
         }

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildService.java
@@ -40,7 +40,6 @@ import org.eclipse.jkube.kit.service.jib.JibServiceUtil;
 import java.io.File;
 import java.io.IOException;
 import java.util.Arrays;
-import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
@@ -109,18 +108,16 @@ public class JibBuildService extends AbstractImageBuildService {
     }
 
     @Override
-    public void push(Collection<ImageConfiguration> imageConfigs, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException {
+    protected void pushSingleImage(ImageConfiguration imageConfiguration, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException {
         try {
-            for (ImageConfiguration imageConfiguration : imageConfigs) {
-                prependRegistry(imageConfiguration, registryConfig.getRegistry());
-                log.info("This push refers to: %s", imageConfiguration.getName());
-                JibServiceUtil.jibPush(
-                    imageConfiguration,
-                    getRegistryCredentials(registryConfig, true, imageConfiguration, log),
-                    getBuildTarArchive(imageConfiguration, configuration),
-                    log
-                );
-            }
+            prependRegistry(imageConfiguration, registryConfig.getRegistry());
+            log.info("This push refers to: %s", imageConfiguration.getName());
+            JibServiceUtil.jibPush(
+                imageConfiguration,
+                getRegistryCredentials(registryConfig, true, imageConfiguration, log),
+                getBuildTarArchive(imageConfiguration, configuration),
+                log
+            );
         } catch (Exception ex) {
             throw new JKubeServiceException("Error when push JIB image", ex);
         }

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildService.java
@@ -16,7 +16,6 @@ package org.eclipse.jkube.kit.config.service.openshift;
 import java.io.File;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
-import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -174,6 +173,12 @@ public class OpenshiftBuildService extends AbstractImageBuildService {
         }
     }
 
+    @Override
+    protected void pushSingleImage(ImageConfiguration imageConfiguration, int retries, RegistryConfig registryConfig, boolean skipTag) {
+        // Do nothing. Image is pushed as part of build phase
+        log.warn("Image is pushed to OpenShift's internal registry during oc:build goal. Skipping...");
+    }
+
     private void applyBuild(String buildName, File dockerTar, KubernetesListBuilder builder)
             throws Exception {
         applyResourceObjects(buildServiceConfig, client, builder);
@@ -183,12 +188,6 @@ public class OpenshiftBuildService extends AbstractImageBuildService {
 
         // Wait until the build finishes
         waitForOpenShiftBuildToComplete(client, build);
-    }
-
-    @Override
-    public void push(Collection<ImageConfiguration> imageConfigs, int retries, RegistryConfig registryConfig, boolean skipTag) throws JKubeServiceException {
-        // Do nothing. Image is pushed as part of build phase
-        log.warn("Image is pushed to OpenShift's internal registry during oc:build goal. Skipping...");
     }
 
     private File getImageStreamFile() {

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/DockerBuildServiceTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/DockerBuildServiceTest.java
@@ -51,6 +51,8 @@ public class DockerBuildServiceTest {
 
   private ImageConfiguration image;
 
+  private ImageConfiguration imageWithSkipEnabled;
+
   @Before
   public void setUp() {
     when(mockedJKubeServiceHub.getDockerServiceHub().getBuildService()).thenReturn(mockedDockerBuildService);
@@ -60,6 +62,12 @@ public class DockerBuildServiceTest {
             .from("from")
             .build()
         ).build();
+    imageWithSkipEnabled = ImageConfiguration.builder()
+        .name("image-name")
+        .build(BuildConfiguration.builder()
+            .skip(true)
+            .build())
+        .build();
   }
 
   @Test
@@ -75,16 +83,8 @@ public class DockerBuildServiceTest {
 
   @Test
   public void build_withImageBuildConfigurationSkipEnabled_shouldNotBuildAndTag() throws Exception {
-    // Given
-    image = ImageConfiguration.builder()
-        .name("image-name")
-        .build(BuildConfiguration.builder()
-            .skip(true)
-            .build())
-        .build();
-
     // When
-    new DockerBuildService(mockedJKubeServiceHub).build(image);
+    new DockerBuildService(mockedJKubeServiceHub).build(imageWithSkipEnabled);
     // Then
     verify(mockedDockerBuildService, times(0))
         .buildImage(eq(image), any(), any());
@@ -108,7 +108,16 @@ public class DockerBuildServiceTest {
     // When
     new DockerBuildService(mockedJKubeServiceHub).push(Collections.emptyList(), 0, null, false);
     // Then
-    verify(mockedJKubeServiceHub.getDockerServiceHub().getRegistryService(), times(1))
-        .pushImages(eq(Collections.emptyList()), eq(0), isNull(), eq(false));
+    verify(mockedJKubeServiceHub.getDockerServiceHub().getRegistryService(), times(0))
+        .pushImage(any(), eq(0), isNull(), eq(false));
+  }
+
+  @Test
+  public void push_withImageBuildConfigurationSkipEnabled_shouldNotPush() throws Exception {
+    // When
+    new DockerBuildService(mockedJKubeServiceHub).build(imageWithSkipEnabled);
+    // Then
+    verify(mockedJKubeServiceHub.getDockerServiceHub().getRegistryService(), times(0))
+        .pushImage(any(), eq(0), isNull(), eq(false));
   }
 }

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildServiceTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/JibBuildServiceTest.java
@@ -197,6 +197,22 @@ public class JibBuildServiceTest {
     }
 
     @Test
+    public void push_withImageBuildConfigurationSkipTrue_shouldNotPushImage() throws JKubeServiceException {
+        // Given
+        imageConfiguration = ImageConfiguration.builder()
+            .name("test/foo:latest")
+            .build(BuildConfiguration.builder()
+                .from("test/base:latest")
+                .skip(true)
+                .build())
+            .build();
+        // When
+        new JibBuildService(mockedServiceHub).push(Collections.singletonList(imageConfiguration), 1, registryConfig, false);
+        // Then
+        jibServiceUtilMockedStatic.verify(() -> JibServiceUtil.jibPush(any(), any(), any(), any()), times(0));
+    }
+
+    @Test
     public void build_withImageBuildConfigurationSkipTrue_shouldNotBuildImage() throws JKubeServiceException {
         // Given
         imageConfiguration = ImageConfiguration.builder()

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenShiftBuildServiceTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenShiftBuildServiceTest.java
@@ -20,6 +20,7 @@ import org.eclipse.jkube.kit.config.image.ImageConfiguration;
 import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
 import org.eclipse.jkube.kit.config.service.JKubeServiceException;
 import org.eclipse.jkube.kit.config.service.JKubeServiceHub;
+import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.mockito.Answers;
@@ -48,13 +49,48 @@ public class OpenShiftBuildServiceTest {
   @Mock(answer = Answers.RETURNS_DEEP_STUBS)
   private KitLogger mockedKitLogger;
 
+  private ImageConfiguration imageConfiguration;
+
+  private ImageConfiguration imageConfigurationWithSkipEnabled;
+
+  @Before
+  public void setUp() {
+    //  @formatter:off
+    imageConfiguration = ImageConfiguration.builder()
+        .name("foo/bar:latest")
+        .build(BuildConfiguration.builder()
+            .from("baseimage:latest")
+            .build())
+        .build();
+    imageConfigurationWithSkipEnabled = ImageConfiguration.builder()
+        .name("foo/bar:latest")
+        .build(BuildConfiguration.builder()
+            .from("baseimage:latest")
+            .skip(true)
+            .build())
+        .build();
+    // @formatter:on
+
+  }
+
   @Test
-  public void push_withDefaults_shouldLogWarning() throws JKubeServiceException {
+  public void push_withEmptyList_shouldNotLogWarning() throws JKubeServiceException {
     // Given
     when(jKubeServiceHub.getLog()).thenReturn(mockedKitLogger);
 
     // When
     new OpenshiftBuildService(jKubeServiceHub).push(Collections.emptyList(), 0, new RegistryConfig(), false);
+    // Then
+    verify(mockedKitLogger, times(0)).warn("Image is pushed to OpenShift's internal registry during oc:build goal. Skipping...");
+  }
+
+  @Test
+  public void push_withValidImage_shouldLogWarning() throws JKubeServiceException {
+    // Given
+    when(jKubeServiceHub.getLog()).thenReturn(mockedKitLogger);
+
+    // When
+    new OpenshiftBuildService(jKubeServiceHub).push(Collections.singletonList(imageConfiguration), 0, new RegistryConfig(), false);
     // Then
     verify(mockedKitLogger, times(1)).warn("Image is pushed to OpenShift's internal registry during oc:build goal. Skipping...");
   }
@@ -62,14 +98,6 @@ public class OpenShiftBuildServiceTest {
   @Test
   public void initClient_withNoOpenShift_shouldThrowException() {
     // Given
-    //  @formatter:off
-    ImageConfiguration imageConfiguration = ImageConfiguration.builder()
-      .name("foo/bar:latest")
-      .build(BuildConfiguration.builder()
-        .from("baseimage:latest")
-        .build())
-      .build();
-    // @formatter:on
     OpenshiftBuildService openshiftBuildService = new OpenshiftBuildService(jKubeServiceHub);
 
     // When + Then
@@ -81,19 +109,12 @@ public class OpenShiftBuildServiceTest {
   @Test
   public void build_withImageBuildConfigurationSkipEnabled_shouldNotBuildImage() throws JKubeServiceException, IOException {
     // Given
-    ImageConfiguration imageConfiguration = ImageConfiguration.builder()
-        .name("foo/bar:latest")
-        .build(BuildConfiguration.builder()
-            .from("baseimage:latest")
-            .skip(true)
-            .build())
-        .build();
     ArchiveService mockedArchiveService = mock(ArchiveService.class, RETURNS_DEEP_STUBS);
     when(jKubeServiceHub.getDockerServiceHub().getArchiveService()).thenReturn(mockedArchiveService);
     OpenshiftBuildService openshiftBuildService = new OpenshiftBuildService(jKubeServiceHub);
 
     // When
-    openshiftBuildService.build(imageConfiguration);
+    openshiftBuildService.build(imageConfigurationWithSkipEnabled);
 
     // Then
     verify(mockedArchiveService, times(0))

